src/crystalla/ndarray.cr

```diff
@@ -0,0 +1,523 @@
+require "./ndarray/builders"
+require "./transformations/pca"
+require "./utils/random_gaussian"
+require "./ndarray/functions"
+
+module Crystalla
+
+
+  # class Dtype(T)
+
+  #   def initialize
+  #   end
+
+  # end
+
+  class Ndarray
+    include LapackHelper
+    include BlasHelper
+
+    getter values
+    getter shape
+    setter shape
+
+    @shape : Tuple(Int32, Int32)
+    @values : Array(Float64)
+
+    def initialize(values, shape = {0,0})
+      @shape = shape
+      @shape = set_shape(values) unless shape != {0,0}
+      @values = set_values(values)
+      #TODO
+      #@dtype = set_type(values)
+    end
+
+
+    def self.[](*values : Array(Number)) : Ndarray
+      Ndarray.new(values)
+    end
+
+    def set_shape(values)
+      if values.is_a?(Array(Int32)) ||values.is_a?(Array(Float64))
+        return {0, values.size}
+      else
+        number_of_cols = values.min_by{|v| v.size}.size
+        values.each{|v| raise ArgumentError.new("Array #{v} should be of dimension #{number_of_cols}") if v.size != number_of_cols}
+        return {values.size, number_of_cols}
+      end
+    end
+
+    def set_values(values)
+      new_values = [] of Float64
+
+      if (values.is_a?(Array(Int32)) ||values.is_a?(Array(Float64)))  # check for 1D array and convert it to Float64 if Int32
+        new_values = values.map{|v| v.to_f64}
+      else  # check for 2D array and store it in column major order
+        (0...shape[1]).each do |i|
+          values.each do |v|
+            new_values << v[i].to_f64
+          end
+        end
+        return new_values
+      end
+    end
+
+    def broadcast(number) : Ndarray
```

Comment:
Just out of curiosity: did you have any reason to annotate all these return types or was it just that you ported them over from the `Matrix` implementation? I think annotating all return types in `Matrix` was a bad idea, because in Crystal that makes us miss some opportunities to take advantage of polymorphism. That was an insight after coding in Crystal for a while after this :). 

Either way, it's ok for now. I guess later on I'll compare this with the `Matrix` implementations and try to extract common code to a module.